src/avk.cpp

```diff
@@ -7101,11 +7103,78 @@ namespace avk
 		return create_semaphore(device(), dispatch_loader_core(), std::move(aAlterConfigBeforeCreation));
 	}
 
+	semaphore root::create_timeline_semaphore(uint64_t aPayload, std::function<void(semaphore_t&)> aAlterConfigBeforeCreation)
+	{
+		return create_semaphore(device(), dispatch_loader_core(), [otherAlterations = move(aAlterConfigBeforeCreation), aPayload](semaphore_t& aSem) {
+			auto typeInfo = std::make_unique<vk::SemaphoreTypeCreateInfo>();
+			typeInfo->semaphoreType = vk::SemaphoreType::eTimeline;
+			typeInfo->initialValue = aPayload;
+
+			aSem.create_info().pNext = typeInfo.get();
+			aSem.set_custom_deleter([aTypeInfo = move(typeInfo)]() { /* Do nothing ... this lambda just keeps the typeInfo struct alive */ });
+			// maybe extend any_owning_resource_t and use handle_lifetime_of instead?
+
+			if (otherAlterations) {
+				otherAlterations(aSem);
+			}
+		});
+	}
+
 	semaphore_t& semaphore_t::handle_lifetime_of(any_owning_resource_t aResource)
 	{
 		mLifetimeHandledResources.push_back(std::move(aResource));
 		return *this;
 	}
+
+	const uint64_t semaphore_t::query_current_value() const {
+		uint64_t value;
+		auto result = mSemaphore.getOwner().getSemaphoreCounterValue(mSemaphore.get(), &value);
+		assert(static_cast<VkResult>(result) >= 0);
+		return value;
+	}
+
+	void semaphore_t::signal(uint64_t aNewValue) const {
+
+		vk::SemaphoreSignalInfo info{};
+		info.sType = vk::StructureType::eSemaphoreSignalInfo;
+		info.pNext = NULL;
+		info.semaphore = mSemaphore.get();
+		info.value = aNewValue;
+
+		mSemaphore.getOwner().signalSemaphore(info);
+	}
+
+	vk::Result semaphore_t::wait_until_signalled(uint64_t aRequiredValue, std::optional<uint64_t> aTimeout) const {
+		return wait_until_signalled({ this }, {aRequiredValue} , true, aTimeout);	// maybe avoid unnecessary vector construction by just creating SemaphoreWaitInfo in this function
+	}
+
+	vk::Result semaphore_t::wait_until_signalled(const std::vector<const semaphore_t*>& aSemaphores, const std::vector<uint64_t>& aTimestamps, bool aWaitOnAll, std::optional<uint64_t> aTimeout) {
+		assert(aSemaphores.size() == aTimestamps.size());
+		if (aSemaphores.size() == 0) {
+			return vk::Result::eSuccess;
+		}
+
+		std::vector<const vk::Semaphore*> semaphores;
+		std::transform(aSemaphores.begin(), aSemaphores.end(), std::back_inserter(semaphores), [](const semaphore_t* s) {return &s->mSemaphore.get(); });
+
+		vk::SemaphoreWaitInfo info{};
+		info.sType = vk::StructureType::eSemaphoreWaitInfo;
+		info.pNext = NULL;
+		info.flags = aWaitOnAll ? vk::SemaphoreWaitFlags() : vk::SemaphoreWaitFlagBits::eAny;
+		info.semaphoreCount = uint32_t(semaphores.size());

```

Comment:
```suggestion
		info.semaphoreCount = static_cast<uint32_T>(semaphores.size());
```

Please use `static_cast` instead! This is C-style.

Reply:
✅